lib/srv/db/common/srvconnect/srvconnect.go

```diff
@@ -0,0 +1,299 @@
+		Certificates: []tls.Certificate{cert},
+		RootCAs:      pool,
+	}, nil
+}
+
+// ShuffleFunc defines a function that shuffles a list of database servers.
+type ShuffleFunc func([]types.DatabaseServer) []types.DatabaseServer
+
+// ShuffleSort is a ShuffleFunc that sorts database servers by name and host ID.
+// Used to provide predictable behavior in tests.
+func ShuffleSort(servers []types.DatabaseServer) []types.DatabaseServer {
+	sort.Sort(types.DatabaseServers(servers))
+	return servers
+}
+
+// ShuffleRandom is a ShuffleFunc that randomizes the order of database servers.
+// Used to provide load balancing behavior when proxying to multiple agents.
+func ShuffleRandom(servers []types.DatabaseServer) []types.DatabaseServer {
+	rand.New(rand.NewSource(time.Now().UnixNano())).Shuffle(
+		len(servers), func(i, j int) {
+			servers[i], servers[j] = servers[j], servers[i]
+		})
+	return servers
+}
+
+type Dialer interface {
+	// Dial dials any address within the site network, in terminating
+	// mode it uses local instance of forwarding server to terminate
+	// and record the connection.
+	Dial(params reversetunnelclient.DialParams) (conn net.Conn, err error)
+}
+
+// ConnectParams contains parameters for connecting to the database server.
+type ConnectParams struct {
+	Logger *slog.Logger
+	// Identity contains the identity information.
+	Identity tlsca.Identity
+	// Servers is the list of database servers that can handle the connection.
+	Servers []types.DatabaseServer
+	// ShuffleFunc is a function used to shuffle the list of database servers.
+	ShuffleFunc ShuffleFunc
+	// Cluster represents the cluster to which the database belongs.
+	Cluster Cluster
+	// Cluster represents the cluster to which the database belongs.
+	Dialer Dialer
+	// CertSigner is used to sign certificates for authenticating with the
+	// database.
+	CertSigner DatabaseCertificateSigner
+	// AuthPreference provides the authentication preferences for the cluster.
+	AuthPreference AuthPreferenceGetter
+	// ClientSrcAddr is the source address of the client making the connection.
+	ClientSrcAddr net.Addr
+	// ClientDstAddr is the destination address of the client making the
+	// connection.
+	ClientDstAddr net.Addr
+}
+
+// ConnectStats contains statistics about the connection attempts.
+type ConnectStats struct {
+	// AttemptedServers is the number of database servers that were attempted to
+	// connect to.
+	AttemptedServers int
+	// DialAttempts is the number of times a dial to a server was attempted.
+	DialAttempts int
+	// DialFailures is the number of times a dial to a server failed.
+	DialFailures int
+}
+
+// Connect connects to the database server running on a remote cluster
+// over reverse tunnel and upgrades this end of the connection to TLS so
+// the identity can be passed over it.
+func Connect(ctx context.Context, params ConnectParams) (net.Conn, ConnectStats, error) {
+	if params.ShuffleFunc == nil {
+		params.ShuffleFunc = ShuffleRandom
+	}
```

Comment:
move to CheckAndSetDefaults? and please do basic safety checks on other parameters.